Build.bee.cs

```diff
@@ -0,0 +1,377 @@
+        {
+            Sdk = sdk;
+            CodeGen = codeGen;
+        }
+
+        public string GetTargetName(string separator = "_")
+        {
+            var targetNameParts = new List<string> {
+                Sdk.Platform.DisplayName.ToLower(),
+                Sdk.Architecture.DisplayName
+            };
+
+            if (CodeGen == CodeGen.Debug)
+            {
+                targetNameParts.Add("dbg");
+            }
+
+            return string.Join(separator, targetNameParts);
+        }
+
+        public string GetArtifactName(string separator = "_")
+        {
+            var artifactNameParts = new List<string> {
+                "tbb",
+                GetTargetName(separator: separator)
+            };
+
+            return string.Join(separator, artifactNameParts);
+        }
+
+        public Sdk Sdk { get; private set; }
+        public CodeGen CodeGen { get; private set; }
+    }
+
+    static void SetupLib(ProgramConfiguration programConfiguration)
+    {
+        var cfg = CfgFor(programConfiguration.CodeGen);
+        var targetId = programConfiguration.GetTargetName();
+
+        NPath[] inputs;
+        var commandLineArguments = new List<string>
+        {
+            $"arch={ArchFor(programConfiguration.Sdk.Architecture)}",
+            $"compiler={CompilerFor(programConfiguration.Sdk)}",
+            $"cfg={cfg}",
+            $"tbb_build_prefix={targetId}",
+            $"-j {Environment.ProcessorCount}",
+        };
+        Dictionary<string, string> environmentVariables;
+
+        var path = Environment.GetEnvironmentVariable("PATH")?.Split(Path.PathSeparator).ToNPaths().ToArray() ?? Array.Empty<NPath>();
+    
+        if (programConfiguration.Sdk is WindowsSdk)
+        {
+            WindowsSdk windowsSdk = (WindowsSdk)programConfiguration.Sdk;
+
+            var binPaths = windowsSdk.EnvironmentVariables["PATH"].Split(";").ToNPaths();
+            inputs = binPaths
+                .Concat(windowsSdk.IncludePaths)
+                .Concat(windowsSdk.LibraryPaths)
+                .ToArray();
+
+            var compileFlags = new[] {
+                "/D_ITERATOR_DEBUG_LEVEL=0"
+            };
+
+            commandLineArguments.Add($"CXXFLAGS=\"{string.Join(" ", compileFlags)}\"");
+
+            environmentVariables = new Dictionary<string, string> {
+                {"PATH", binPaths
+                    .Concat(path)
+                    .Select(p => p.ResolveWithFileSystem().MakeAbsolute().ToString(SlashMode.Native))
+                    .SeparateWith(";")},
+                {"INCLUDE", windowsSdk.IncludePaths
+                    .Select(p => p.ResolveWithFileSystem().MakeAbsolute().ToString(SlashMode.Native))
+                    .SeparateWith(";")},
+                {"LIB", windowsSdk.LibraryPaths
+                    .Select(p => p.ResolveWithFileSystem().MakeAbsolute().ToString(SlashMode.Native))
+                    .SeparateWith(";")}
```

Comment:
In general we want to avoid using absolute paths in compiler/linker commands, because it makes the build vary more between machines. Are you sure this is needed? (And similarly for the other `MakeAbsolute` calls in this file.)

Reply:
This was needed, although I can't remember why. I think the make build system changes directory and the relative environment variables no longer work but I can double check and add a comment.

When you say it makes the build vary between machines, do you mean that the build graph will be different or that the build will be non-deterministic? Either way I don't think this is of particular concern to us at this time.

Reply:
Generally make/cmake likes to change working directory mid build. You can't get away with relative paths.

Reply:
Ah, recursive makefiles. OK.

> When you say it makes the build vary between machines, do you mean that the build graph will be different or that the build will be non-deterministic? Either way I don't think this is of particular concern to us at this time.

If you use absolute paths then the shape of the build graph is the same, but the values of absolute paths used differ when the same code is built in a different directory. Then things like whitespace and non-ascii characters in the checkout path can cause the build to fail on some checkout locations but not others (eg if the build is missing some quotes). It also prevents cross-developer cache hits for the same code built from different directories if you use a network cache. And makes it easier to hit annoying windows path length limits.

In an ideal setup, the build should work the same regardless of the directory it's checked out in. But this sounds like something we have to live with in this case.